File: LargestInRow.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null)
            return result;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int max = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            max = Integer.MIN_VALUE;
            for (int i = 0; i < size; i++) {
                TreeNode curr = q.poll();
                max = Math.max(max, curr.val);
                if (curr.left != null) {
                    q.add(curr.left);
                }
                if (curr.right != null) {
                    q.add(curr.right);
                }
            }
            result.add(max);
        }
        return result;
    }

    //DFS O(n) O(h)
    // List<Integer> result;
    // public List<Integer> largestValues(TreeNode root) {
    //     result= new ArrayList<>();
    //     dfs(root,0);
    //     return result;
    // }
    // private void dfs(TreeNode root,int level){
    //     if(root==null) return;
    //     if(result.size()==level){
    //         result.add(root.val);
    //     }
    //     result.set(level,Math.max(root.val,result.get(level)));
    //     dfs(root.left,level+1);
    //     dfs(root.right,level+1);
    // }
}

File: largest in row.py
# TC = O(n)
# SC = O(n)
# run on LC

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        output = []
        q = []
        q.append(root)

        if not root:
            return []
        while (q):
            ls = len(q)
            ans = []
            for i in range(ls):
                node = q.pop(0)
                ans.append(node.val)

                if (node.left != None):
                    q.append(node.left)
                if (node.right != None):
                    q.append(node.right)
            output.append(max(ans))

        return output
